File: app/src/main/java/io/github/zwieback/familyfinance/core/preference/custom/BooleanPreference.java
package io.github.zwieback.familyfinance.core.preference.custom;

import android.content.Context;
import android.util.AttributeSet;
import androidx.annotation.CallSuper;
import androidx.preference.CheckBoxPreference;
import androidx.preference.Preference;

public abstract class BooleanPreference
        extends CheckBoxPreference
        implements Preference.OnPreferenceChangeListener {

    @SuppressWarnings("unused")
    public BooleanPreference(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        init(context);
    }

    @SuppressWarnings("unused")
    public BooleanPreference(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context);
    }

    @SuppressWarnings("unused")
    public BooleanPreference(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context);
    }

    @SuppressWarnings("unused")
    public BooleanPreference(Context context) {
        super(context);
        init(context);
    }

    @CallSuper
    protected void init(Context context) {
        setOnPreferenceChangeListener(this);
    }
}

File: app/src/main/java/io/github/zwieback/familyfinance/core/preference/custom/BooleanPreference.kt
package io.github.zwieback.familyfinance.core.preference.custom

import android.content.Context
import android.util.AttributeSet
import androidx.core.content.res.TypedArrayUtils
import androidx.preference.CheckBoxPreference
import androidx.preference.Preference
import io.github.zwieback.familyfinance.R
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlin.coroutines.CoroutineContext

abstract class BooleanPreference @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = TypedArrayUtils.getAttr(
        context,
        R.attr.checkBoxPreferenceStyle,
        android.R.attr.checkBoxPreferenceStyle
    ),
    defStyleRes: Int = 0
) : CheckBoxPreference(context, attrs, defStyleAttr, defStyleRes),
    Preference.OnPreferenceChangeListener,
    CoroutineScope {

    private lateinit var rootJob: Job

    override val coroutineContext: CoroutineContext
        get() = Dispatchers.Main + rootJob

    override fun onAttached() {
        super.onAttached()
        rootJob = Job()
        onPreferenceChangeListener = this
    }

    override fun onDetached() {
        rootJob.cancel()
        super.onDetached()
    }
}
